### ISSUE TITLE
Handling of redirects

### ISSUE THREAD

**Issue by reversepanda** [bustoutsolutions/siesta#210]
Hi there,

I wonder if the only way to handle redirects with Siesta is to provide our own NetworkProvider? Or is that possible at all?

We would either like to be able to enable/disable redirects for our current usage case, or (even) better have Siesta catch redirects and apply any configurations to the redirect request before sending. I'm not sure if this is feasible.

Any hints how to handle such a case?

Sincerely,
Kaweh

**Comment by pcantrell** [comment]
Right now, Siesta leaves all redirect handling to the network provider. Any custom redirect handling would belong in a `URLSession` with a custom delegate, a custom `taskWillPerformHTTPRedirection` handler with Alamofire, etc.

I didn't surface redirect handling in Siesta because it wasn't clear to me what the use cases are, i.e. what problem Siesta should solve that a network provider shouldn’t. In particular, it's not clear to me that there's a good general policy that handles the security implications of following or not following redirects from an API, to the same host or not, with or without the same headers sent. Thus [this note to self](https://github.com/bustoutsolutions/siesta/blob/baff44dabd291d58e38094614a894262c262a502/Tests/Functional/ResourceStateSpec.swift#L298).

If you can lay out your particular problem in a little more detail, we might be able to spot some good abstraction for Siesta to provide. Hearing about real-world use cases is always very helpful.

**Comment by pcantrell** [comment]
@reversepand Ping on this issue: would you be willing to describe your use case a little? What kind of redirects does your API send, and what is your desired policy for handling them?

The network provider may well be the right solution, but I'd like to understand your need to make sure that Siesta is doing what it ought to be doing.

**Comment by reversepanda** [comment]
@pcantrell Sorry for coming back that late.

Our concrete use case was related to our backend which uses a OAuth token. The OAuth token is added to all requests via decorating (very much like described in the docs). We did a POST request that resulted in a redirect to the final resource, which would fail due to bad auth. Naturally the redirect wouldn't be decorated with the token, because Siesta would never see the redirect.

Now I understand this is a very edgy use case.

To solve this we just disabled redirects via setting the URLSession delegate (as you already recommended):

```swift
class OurService {
       ...

        func urlSession(_ session: URLSession,
                        task: URLSessionTask,
                        willPerformHTTPRedirection response: HTTPURLResponse,
                        newRequest request: URLRequest,
                        completionHandler: @escaping (URLRequest?) -> Void) {
            completionHandler(nil)
        }

        ...

            let urlSession = URLSession(configuration: .ephemeral, delegate: self, delegateQueue: nil)
            service = Service(baseURL: "https://....", networking: urlSession)
        ...
}
```

We would just chain the redirect request manually (which was the right way to do anyways in our case, because we would need the "intermediate" result).

PS: we got back to our backend team and asked them to change the API and replace the redirect 30x with a 20x, because the API design was suboptimal. So in the end this resolved itself anyways.

Thus I believe you are correct and such use cases can be easily handled by app code without direct support by Siesta. I wonder about redirects with REST APIs myself and I'm not sure if they are very useful. Redirects change the flow of data and influence things like app entity caches etc. and we would lose control.

The only way I could think of adding support in Siesta, would probably mean a lot of code in Siesta to catch all redirects and send them through the configure process to allow the app code to "change the redirect" as needed to handle similar use cases. But I'm not sure if it's worth the complexity.

In conclusion: there is an easy solution (maybe worth adding explicitly to the documentation?) and you can better judge if the added complexity to catch redirect requests, configure them and resend them via Siesta is worth the effort. For now we don't have a need for this and are happy to use Siesta as is (which is a wonderful framework).

**Comment by nabbestemmia** [comment]
Thank you @reversepanda 
This topic was really useful to me.
In my humble opinion, redirects should be banned from REST APIs.
May I ask you how changing from 30x to 20x affected the behavior of your application?

**Comment by reversepanda** [comment]
Hi @nabbestemmia, the change didn't affect us a lot. It essentially removed described workaround in the client code to avoid auto-redirects. 

I agree, It's a API design issue as far as we were concerned. The redirect was there to chain the POST to the final resource and the backend team believed this would be easier for the client to handle, which was not the case. 

We rather preferred explicit vs. implicit and now do a POST followed by a GET to get what we want, allowing us setup the request independently with all the needed custom headers (which a unattended redirect would drop) and also to introspect the responses independently, which was also important in our specific case.

**Comment by nabbestemmia** [comment]
@reversepanda 
Similar scenario here.
Thank you for the infos!
